pkg/persistence/redis/store.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/goharbor/harbor-scanner-clair/pkg/job"
+	"github.com/goharbor/harbor-scanner-clair/pkg/persistence"
+	"github.com/gomodule/redigo/redis"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/xerrors"
+	"time"
+)
+
+type redisStore struct {
+	cfg  etc.StoreConfig
+	pool redis.Pool
+}
+
+func NewStore(cfg etc.StoreConfig) persistence.Store {
+	return &redisStore{
+		cfg: cfg,
+		pool: redis.Pool{
+			Dial: func() (redis.Conn, error) {
+				return redis.DialURL(cfg.RedisURL)
+			},
+			MaxIdle:   cfg.PoolMaxIdle,
+			MaxActive: cfg.PoolMaxActive,
+			Wait:      true,
+		},
+	}
+}
+
+func (rs *redisStore) Save(scanJob job.ScanJob) error {
+	conn := rs.pool.Get()
+	defer rs.close(conn)
+
+	bytes, err := json.Marshal(scanJob)
+	if err != nil {
+		return xerrors.Errorf("marshalling scan job: %w", err)
+	}
+
+	key := rs.getKeyForScanJob(scanJob.ID)
+
+	log.WithFields(log.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"redis_key":       key,
+	}).Debug("Saving scan job")
+
+	_, err = conn.Do("SET", key, string(bytes))
+	if err != nil {
+		return xerrors.Errorf("saving scan job: %w", err)
+	}
+
+	return nil
+}
+
+func (rs *redisStore) saveWithExpire(scanJob job.ScanJob, expire time.Duration) error {
+	conn := rs.pool.Get()
+	defer rs.close(conn)
+
+	bytes, err := json.Marshal(scanJob)
+	if err != nil {
+		return xerrors.Errorf("marshalling scan job: %w", err)
+	}
+
+	key := rs.getKeyForScanJob(scanJob.ID)
+
+	log.WithFields(log.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"expire":          expire.String(),
+		"redis_key":       key,
+	}).Debug("Saving scan job with expire")
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return err
+	}
+	err = conn.Send("SET", key, string(bytes))
+	if err != nil {
+		return err
+	}
+	err = conn.Send("EXPIRE", key, int(expire.Seconds()))
+	if err != nil {
+		return err
+	}
+	_, err = conn.Do("EXEC")
+	if err != nil {
+		return xerrors.Errorf("saving scan job: %w", err)
+	}
+
+	return nil
+}
+
+func (rs *redisStore) Get(scanJobID string) (*job.ScanJob, error) {
+	conn := rs.pool.Get()
+	defer rs.close(conn)
+
+	key := rs.getKeyForScanJob(scanJobID)
+	value, err := redis.String(conn.Do("GET", key))
+	if err != nil {
+		if err == redis.ErrNil {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var scanJob job.ScanJob
+	err = json.Unmarshal([]byte(value), &scanJob)
+	if err != nil {
+		return nil, err
+	}
+
+	return &scanJob, nil
+}
+
+func (rs *redisStore) UpdateStatus(scanJobID string, newStatus job.Status, error ...string) error {
+	log.WithFields(log.Fields{
+		"scan_job_id": scanJobID,
+		"new_status":  newStatus.String(),
+	}).Debug("Updating status for scan job")
+
+	scanJob, err := rs.Get(scanJobID)
+	if err != nil {
+		return err
+	}
+
+	scanJob.Status = newStatus
+	if len(error) > 0 {
+		scanJob.Error = error[0]
+	}
+
+	if newStatus == job.Finished || newStatus == job.Failed {
```

Comment:
I know this does make sense. However, have you considered the case that if the job is crashed and then it will never reach such two status? Dirty data may be generated then.

Maybe exp can be assigned to all of the data.

Reply:
Good point. This condition should never eval to true. However the job might fail and leave the scan job in Pending status. Therefore I set TTL for scan jobs in Pending state as well to auto-remove stale data.